gh-132097: allow AC to disable fastcall convention to avoid UBSan failures

picnixz opened the pull request:
We need a way to explicitly disable fastcall convention for non-new methods. Let me explain why. Consider the following case:

```c
/*[clinic input]
@classmethod
_testclinic.TestClass.__new__ as varpos_no_fastcall
    *args: tuple
[clinic start generated code]*/
```

What clinic generates is two signatures:

```c
PyObject *varpos_no_fastcall(PyTypeObject *type, PyObject *args, PyObject *kwargs);
PyObject *varpos_no_fastcall_impl(PyTypeObject *type, PyObject *args);
```

At runtime, the function being called will be `varpos_no_fastcall`. In particular, we need to make it a `METH_VARARGS | METH_KEYWORD | METH_CLASS`. **However**, and this is where the runtime UB happens, the fact that we are having `METH_VARARGS` in `methodobject.c::cfunction_call()` implies that it will be used as follows:

```c
static PyObject *
cfunction_call(PyObject *func, PyObject *args, PyObject *kwargs)
{
    assert(kwargs == NULL || PyDict_Check(kwargs));

    PyThreadState *tstate = _PyThreadState_GET();
    assert(!_PyErr_Occurred(tstate));

    int flags = PyCFunction_GET_FLAGS(func);
    if (!(flags & METH_VARARGS)) {
        /* If this is not a METH_VARARGS function, delegate to vectorcall */
        return PyVectorcall_Call(func, args, kwargs);
    }

    /* For METH_VARARGS, we cannot use vectorcall as the vectorcall pointer
     * is NULL. This is intentional, since vectorcall would be slower. */
    PyCFunction meth = PyCFunction_GET_FUNCTION(func);
    PyObject *self = PyCFunction_GET_SELF(func);

    PyObject *result;
    if (flags & METH_KEYWORDS) {
        result = _PyCFunctionWithKeywords_TrampolineCall(
            (*(PyCFunctionWithKeywords)(void(*)(void))meth),
            self, args, kwargs);
    }
    else {
        if (kwargs != NULL && PyDict_GET_SIZE(kwargs) != 0) {
            _PyErr_Format(tstate, PyExc_TypeError,
                          "%.200s() takes no keyword arguments",
                          ((PyCFunctionObject*)func)->m_ml->ml_name);
            return NULL;
        }
        result = _PyCFunction_TrampolineCall(meth, self, args);
    }
    return _Py_CheckFunctionResult(tstate, func, result, NULL);
}
```

The important bits are:

```c
    if (flags & METH_KEYWORDS) {
        result = _PyCFunctionWithKeywords_TrampolineCall(
            (*(PyCFunctionWithKeywords)(void(*)(void))meth),
            self, args, kwargs);
    }
```

As you can see, `self` is being passed as a `PyObject *` to the called method. But, here, `meth` is `varpos_no_fastcall`, which expects a `PyTypeObject *` instead, hence the undefined behaviour. Now, the unfortunate part is that we *must* use all flags `METH_VARARGS | METH_KEYWORD | METH_CLASS` here.

One could wonder: why not using something else as a base classmethod instead of `__new__`? well.. the answer is that if a classmethod that is not `__new__` is generated by AC, then that method will use a `METH_FASTCALL` convention, which **entirely defeats the purpose of the test**.

For efficiency purposes (and that's what we want obviously), if we use 

```c
/*[clinic input]
@classmethod
_testclinic.TestClass.varpos_no_fastcall as varpos_no_fastcall
    *args: tuple
[clinic start generated code]*/
```

AC will generate:

```c
PyObject *varpos_no_fastcall(PyObject *type, PyObject *const *args, Py_ssize_t nargs)
PyObject *varpos_no_fastcall_impl(PyObject *type, PyObject *args)
```

and we will also have some autogenerated `METHODDEF` macro containing `METH_FASTCALL|METH_CLASS` flags. We also cannot just use `varpos_no_fastcall` without the `METH_FASTCALL` flag as its signature wouldn't be compatible for `cfunction_call` which expects either of the two forms:

```c
PyObject *meth(PyObject *self, PyObject *args);
PyObject *meth(PyObject *self, PyObject *args, PyObject *kwargs);
```

So the solution I came up with is a generic `@disable` directive that can disable the fastcall convention explicitly even for non-`__new__` class methods:

```c
/*[clinic input]
@disable fastcall
@classmethod
_testclinic.TestClass.varpos_no_fastcall as varpos_no_fastcall
    *args: tuple
[clinic start generated code]*/
```

AC will generate:

```c
PyObject *varpos_no_fastcall(PyObject *type, PyObject *args);
PyObject *varpos_no_fastcall_impl(PyTypeObject *type, PyObject *args);
```

and a `METHODDEF` macro with `METH_VARARGS|METH_CLASS`. Note that now `varpos_no_fastcall` has the expected signature and can be correctly called at runtime by `cfunction_call`. I've explained all of this in the comments of the tested functions (`_testclinic.c`). Since it's part of the tests, I think we can safely keep those comments.

<!-- gh-issue-number: gh-132097 -->
* Issue: gh-132097
<!-- /gh-issue-number -->

picnixz committed:
remove unused casts by clinic

picnixz committed:
fixup

picnixz committed:
Revert "fixup"

This reverts commit 4001c734022088eba7a403dc01fe6645dd3276fe.

picnixz committed:
fixup

picnixz committed:
fixup

picnixz committed:
Merge remote-tracking branch 'upstream/main' into fix/ubsan/clinic-test-111178

picnixz committed:
move comments

picnixz committed:
move comments

vstinner reviewed (commented):
This PR changes two things: remove cast and add `@disable`. It changes 119 files so it's hard to review. You should split your PR int two PRs.

I'm not excited by the removal of the cast since it impacts tons of files (all of them?). I'm +0 on it.

picnixz committed:
Merge remote-tracking branch 'upstream/main' into fix/ubsan/clinic-test-111178

picnixz committed:
revert cast removal

picnixz committed:
make clinic

vstinner commented on Tools/clinic/libclinic/dsl_parser.py:
```diff
@@ -423,6 +425,13 @@ def at_critical_section(self, *args: str) -> None:
+            fail("Can't set @disable, function is not a normal callable")
+        if not args:
+            fail("@disable expects at least one argument")
+        self.disable_fastcall = 'fastcall' in args
```
I would prefer stricter parsing, something like:

```suggestion
        if args == 'fastcall':
            self.disable_fastcall = True
        else:
            self.fail("Invalid @disable option")
```

picnixz replied:
Since `args` is a list of arguments, I don't think I'm able to do it that way. Directives are `@NAME arg1 arg2 ...`. So what I can do is however check that `args` has length 1 only.

picnixz committed:
Merge remote-tracking branch 'upstream/main' into fix/ubsan/clinic-test-111178

picnixz committed:
Merge remote-tracking branch 'upstream/main' into fix/ubsan/clinic-test-111178

markshannon committed:
GH-131498: Cases generator: manage stacks automatically (GH-132074)

picnixz committed:
gh-99108: Implement HACL* HMAC (#130157)

A new extension module, `_hmac`, now exposes the HACL* HMAC (formally verified) implementation.

The HACL* implementation is used as a fallback implementation when the OpenSSL implementation of HMAC
is not available or disabled. For now, only named hash algorithms are recognized and SIMD support provided
by HACL* for the BLAKE2 hash functions is not yet used.

picnixz committed:
check arguments given to the 'disable' directive

picnixz committed:
Merge branch 'main' into fix/ubsan/clinic-test-111178

picnixz committed:
Merge remote-tracking branch 'upstream/main' into fix/ubsan/clinic-test-111178

picnixz committed:
Merge branch 'main' into fix/ubsan/clinic-test-111178

vstinner reviewed (approved):
LGTM
